In [58]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import warnings
warnings.filterwarnings("ignore")
%matplotlib inline

## Load Data

In [59]:
df = pd.read_csv("EasyVisa.csv")
df.head()

case_id continent education_of_employee has_job_experience  \
0  EZYV01      Asia           High School                  N   
1  EZYV02      Asia              Master's                  Y   
2  EZYV03      Asia            Bachelor's                  N   
3  EZYV04      Asia            Bachelor's                  N   
4  EZYV05    Africa              Master's                  Y   

  requires_job_training  no_of_employees  yr_of_estab region_of_employment  \
0                     N            14513         2007                 West   
1                     N             2412         2002            Northeast   
2                     Y            44444         2008                 West   
3                     N               98         1897                 West   
4                     N             1082         2005                South   

   prevailing_wage unit_of_wage full_time_position case_status  
0         592.2029         Hour                  Y      Denied  
1       83425.6500         Year                  Y   Certified  
2      122996.8600         Year                  Y      Denied  
3       83434.0300         Year                  Y      Denied  
4      149907.3900         Year                  Y   Certified

In [60]:
df.shape

(25480, 12)

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25480 entries, 0 to 25479
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   case_id                25480 non-null  object 
 1   continent              25480 non-null  object 
 2   education_of_employee  25480 non-null  object 
 3   has_job_experience     25480 non-null  object 
 4   requires_job_training  25480 non-null  object 
 5   no_of_employees        25480 non-null  int64  
 6   yr_of_estab            25480 non-null  int64  
 7   region_of_employment   25480 non-null  object 
 8   prevailing_wage        25480 non-null  float64
 9   unit_of_wage           25480 non-null  object 
 10  full_time_position     25480 non-null  object 
 11  case_status            25480 non-null  object 
dtypes: float64(1), int64(2), object(9)
memory usage: 2.3+ MB


In [62]:
df.describe(include = 'all').T

count unique         top   freq          mean  \
case_id                  25480  25480      EZYV01      1           NaN   
continent                25480      6        Asia  16861           NaN   
education_of_employee    25480      4  Bachelor's  10234           NaN   
has_job_experience       25480      2           Y  14802           NaN   
requires_job_training    25480      2           N  22525           NaN   
no_of_employees        25480.0    NaN         NaN    NaN    5667.04321   
yr_of_estab            25480.0    NaN         NaN    NaN   1979.409929   
region_of_employment     25480      5   Northeast   7195           NaN   
prevailing_wage        25480.0    NaN         NaN    NaN  74455.814592   
unit_of_wage             25480      4        Year  22962           NaN   
full_time_position       25480      2           Y  22773           NaN   
case_status              25480      2   Certified  17018           NaN   

                                std     min       25%       50%          75%  \
case_id                         NaN     NaN       NaN       NaN          NaN   
continent                       NaN     NaN       NaN       NaN          NaN   
education_of_employee           NaN     NaN       NaN       NaN          NaN   
has_job_experience              NaN     NaN       NaN       NaN          NaN   
requires_job_training           NaN     NaN       NaN       NaN          NaN   
no_of_employees        22877.928848   -26.0    1022.0    2109.0       3504.0   
yr_of_estab               42.366929  1800.0    1976.0    1997.0       2005.0   
region_of_employment            NaN     NaN       NaN       NaN          NaN   
prevailing_wage        52815.942327  2.1367  34015.48  70308.21  107735.5125   
unit_of_wage                    NaN     NaN       NaN       NaN          NaN   
full_time_position              NaN     NaN       NaN       NaN          NaN   
case_status                     NaN     NaN       NaN       NaN          NaN   

                             max  
case_id                      NaN  
continent                    NaN  
education_of_employee        NaN  
has_job_experience           NaN  
requires_job_training        NaN  
no_of_employees         602069.0  
yr_of_estab               2016.0  
region_of_employment         NaN  
prevailing_wage        319210.27  
unit_of_wage                 NaN  
full_time_position           NaN  
case_status                  NaN

#### Observations:
<li>continent: There are 25,480 non-null values, 6 unique continents, with "Asia" being the most frequent one, appearing 16,861 times.
<li>no_of_employees: The average number of employees is around 5,667, but the standard deviation is quite high (22,877), indicating a large spread. The minimum value is -26 (potentially an error or missing value), while the maximum is 602,069.
<li>prevailing_wage: The average wage is 74,455, with a large range from 2 to over 319,000, suggesting significant wage variation.
<li>The large standard deviations in numerical columns indicate the presence of outliers or a wide range of values

In [69]:
df.loc[df['no_of_employees'] < 0]

case_id      continent education_of_employee has_job_experience  \
245      EZYV246         Europe              Master's                  N   
378      EZYV379           Asia            Bachelor's                  N   
832      EZYV833  South America              Master's                  Y   
2918    EZYV2919           Asia              Master's                  Y   
6439    EZYV6440           Asia            Bachelor's                  N   
6634    EZYV6635           Asia            Bachelor's                  Y   
7224    EZYV7225         Europe             Doctorate                  N   
7281    EZYV7282           Asia           High School                  N   
7318    EZYV7319           Asia            Bachelor's                  Y   
7761    EZYV7762           Asia              Master's                  N   
9872    EZYV9873         Europe              Master's                  Y   
11493  EZYV11494           Asia           High School                  Y   
13471  EZYV13472  North America              Master's                  N   
14022  EZYV14023           Asia            Bachelor's                  N   
14146  EZYV14147           Asia            Bachelor's                  N   
14726  EZYV14727           Asia              Master's                  N   
15600  EZYV15601           Asia            Bachelor's                  N   
15859  EZYV15860           Asia           High School                  N   
16157  EZYV16158           Asia              Master's                  Y   
16883  EZYV16884  North America            Bachelor's                  Y   
17006  EZYV17007           Asia             Doctorate                  Y   
17655  EZYV17656  North America            Bachelor's                  Y   
17844  EZYV17845           Asia            Bachelor's                  N   
17983  EZYV17984           Asia            Bachelor's                  N   
20815  EZYV20816           Asia            Bachelor's                  N   
20984  EZYV20985         Europe             Doctorate                  Y   
21255  EZYV21256  North America           High School                  N   
21760  EZYV21761           Asia            Bachelor's                  Y   
21944  EZYV21945         Africa              Master's                  Y   
22084  EZYV22085  North America            Bachelor's                  Y   
22388  EZYV22389           Asia              Master's                  Y   
23186  EZYV23187           Asia              Master's                  N   
23476  EZYV23477         Europe              Master's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
245                       N            -14.0         1980   
378                       Y            -14.0         2011   
832                       N            -14.0         2002   
2918                      N            -14.0         2005   
6439                      N            -14.0         2013   
6634                      N            -14.0         1923   
7224                      N            -14.0         1998   
7281                      N            -14.0         2000   
7318                      Y            -14.0         2006   
7761                      N            -14.0         2009   
9872                      N            -14.0         1996   
11493                     N            -14.0         1999   
13471                     N            -14.0         2003   
14022                     Y            -14.0         1946   
14146                     Y            -14.0         1954   
14726                     N            -14.0         2000   
15600                     N            -14.0         2014   
15859                     N            -14.0         1969   
16157                     N            -14.0         1994   
16883                     N            -14.0         1968   
17006                     N            -14.0         1984   
17655                     N            -14.0         2007   
17844                

In [70]:
df.shape

(25480, 12)

In [68]:
# Calculate the median of non-zero, positive no_of_employees values
median_value = df[df['no_of_employees'] > 0]['no_of_employees'].median()
df['no_of_employees'] = df['no_of_employees'].apply(lambda x: median_value if x < 0 else x)